File: scala/bob/src/main/scala/Bob.scala


object Bob {

  def response(statement: String): String = {
    val trimedStatement = statement.trim
    if (trimedStatement.isEmpty)
      "Fine. Be that way!"
    else if (isYelling(trimedStatement)) {
      if (trimedStatement.endsWith("?"))
        "Calm down, I know what I'm doing!"
      else
        "Whoa, chill out!"
    } else if (trimedStatement.endsWith("?")) {
      "Sure."
    } else {
      "Whatever."
    }
  }

  private[this] def isYelling(statement: String): Boolean = {
    val letters = statement.filter(c => c.isLetter)
    if (letters.isEmpty)
      false
    else
      letters.forall(c => c.isUpper)
  }
}

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Text as T
import Data.Char

responseFor :: String -> String
responseFor xs
    | trimedSentence == T.empty = "Fine. Be that way!"
    | isQuestion && isYelling = "Calm down, I know what I'm doing!"
    | isYelling = "Whoa, chill out!"
    | isQuestion = "Sure."
    | otherwise = "Whatever."

    where
        trimedSentence = T.strip $ pack xs
        chars = T.filter isAlpha trimedSentence
        isYelling = if T.null chars then False else T.all isUpper chars
        lastCharMaybe = T.unsnoc trimedSentence
        isQuestion = case lastCharMaybe of
            Just (_, lastChar) -> lastChar == '?'
            _ -> False